Assets/GRPC.NET/Scripts/GRPCBestHttpHandler.cs

```diff
@@ -149,6 +149,30 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage grpcRe
                 return true;
             };
 
+            bestRequest.Callback += (request, response) =>
+            {
+                // If response IS NOT completed yet there was an exception in the request
+                if (!grpcResponseTask.Task.IsCompleted)
+                {
+                    grpcResponseTask.SetException(request?.Exception);
```

Comment:
If Grpc handler on the server throws an Exception we will get inside this branch but request.Exception will be null. However grpc-status and grpc-message header fields will contain error into so instead of this one line we should probably have this:

```c#
var exception = request?.Exception;
if (exception == null && response != null &&
    response.Headers.TryGetValue("grpc-status", out var strStatus) &&
    int.TryParse(strStatus[0], out var status))
{
	var message = "UnknownError";
	if (response.Headers.TryGetValue("grpc-message", out var messages))
	{
		message = messages[0];
	}

	exception = new RpcException(new Status((StatusCode) status, message));
}

exception ??= new RpcException(new Status(StatusCode.Unknown, "Exception was thrown by handler"));

grpcResponseTask.SetException(exception);
```

Reply:
Hi @igor84 , thanks for highlighting this case. I actually believe the underlaying issue is that in this case the complete response is actually recieved but not handled correctly downstream.

So instead of adding this code and throwing the error by ourself, we should actually make sure it is done by grpc internally (as it should have the grpc-status available there anyway). 
Unfortunatly it looks like we are not forwarding the right grpc-status because the differentiation of leading and trailing headers in `OnHeadersReceived` is implemented in a too naive way and fails if a DATA package was never sent. I found a few additional cases where we should also handle it differently.

I need to figure out how I can do this better, its much more complex then anticipated. And I need to extend my testing code otherwise I will break things left and right.

Reply:
That makes sense. Great find.